# Assignment 1 - Introduction to Python - Loops

## Part 1
* Assign x the value 16
* Compute and print x, x squared, and x cubed.
* Compute x divided by 3 and then multiply the result by 3.  Does the result have the same as value as x.
* Redo the previous parts but assign x the floating point value 5.0

_After the code section, create a MarkDown section and explain your results.  Do you expect the same results if you use // instead of /?_

In [8]:
x = 16
print(x)
print(x ** 2)
print(x ** 3)
print((x/3)*3)
x = 5.0
print(x)
print(x ** 2)
print(x ** 3)
print((x/3)*3)

16
256
4096
16.0
5.0
25.0
125.0
5.0


The results here show that if you are divide an integer by an integer using / it will return a floating point number because it is willing to give an exact value. If you were to use // then it would return 15 for the first and 3.0 for the second because it divides the number and then rounds down to the nearest integer.

## Part 2
* Ask the user for a string and assign it to `scores`
* Split `scores` into a list
* Iterate over the list and count the number of words that start with "0", contain "4", and meet neither condition. Print the results.
* Iterate over the list and compute the sum of the values that are positive.


In [23]:
scores = input("Enter your scores:")
count0 = 0
count4 = 0
count = 0
scores = scores.split()
print(scores)
for x in scores :
    if x.startswith("0") == True :
        count0 += 1
        if "4" in x :
            count4 += 1
    else :
        count += 1
print("Scores that start with 0: " + str(count0))
print("Scores that contain 4: " + str(count4))
print("Scores with neither: " + str(count))

total = 0
for x in scores :
    if int(x) > 0 :
       total += int(x)
print(total)

        

Enter your scores: 10 20 30 40 50 60 70 84 01 02 03 60 04


['10', '20', '30', '40', '50', '60', '70', '84', '01', '02', '03', '60', '04']
Scores that start with 0: 4
Scores that contain 4: 1
Scores with neither: 9
434


## Part 3
* Use a for loop to compute the value of 1000 to the power (n-1)/n where n starts at 2 and ends at 15. 

In [25]:
for x in range(2,15) :
    print(1000 ** ((x-1)/x))


31.622776601683793
99.99999999999997
177.82794100389228
251.1886431509581
316.227766016838
372.7593720314939
421.6965034285823
464.15888336127773
501.18723362727235
533.6699231206309
562.3413251903489
587.8016072274914
610.540229658533


## Part 4
* Ask the user for a numeric value x.
* Use a while loop to compute and print a sequence of values where the new value is based on the old value and uses the following formula: $x_{new} = x_{old} - (x_{old}^2-7)/(2x_{old})$
* Continue iterating as long as the absolute value of the difference between the old and new values is greater than 0.000001.  _Use the abs() function._
* After the code section, create a MarkDown section and give the results of trying different starting values of x, some suggestions are:

    * 0, 1, 2, 2.5, 2.6, 3, 5, 10, 20, 100
    * Negative values
    * Do you always get the same result?

Note: This will compute the square root of 7 using the Newton-Raphson method.  It requires Calculus, but if you want to learn about finding zeros of a function it is a good place to start. Also, it is relevant to techniques that we will use later in the course.


    

In [47]:
x = input("Give a numeric value:")
x = float(x)
newx = x - (x ** 2 -7)/(2 * x)
while abs(x - newx) > .000001 :
    temp = newx
    print("Old x: " + str(x) + " New x: " + str(newx))
    x = temp
    newx = x - (x ** 2 - 7) / (2 * x)
print("Old x: " + str(x) + " New x: " + str(newx))

Give a numeric value: 14.6


Old x: 14.6 New x: 7.5397260273972595
Old x: 7.5397260273972595 New x: 4.234070862535839
Old x: 4.234070862535839 New x: 2.943663070160058
Old x: 2.943663070160058 New x: 2.6608263067573774
Old x: 2.6608263067573774 New x: 2.645794014997305
Old x: 2.645794014997305 New x: 2.645751311409218
Old x: 2.645751311409218 New x: 2.6457513110645907


The different results:
* 0 gave a divide by 0 error
* 1 ran through 5 times with the final numbers matching at 2.6457513
* 2 ran through 3 times with the final numbers matching at 2.6457513
* 50 ran through 8 times with the final numbers matching at 2.6457513
* 100 ran through 9 times with the final numbers matching at 2.6457513
* 14.6 ran through 6 times with the final numbers matching at 2.6457513

## Bonus
Suppose we are at a ice cream store that serves 31 flavors.  We are going to choose `r` scoops where the order of selection does not matter and you can select a flavor more than once. The number of ways that we can do this is $C(r+30,r)$ where $C(n,k) = {n!\over{(n-k)! \times k!}}$.

  For example, If you are going to order 10 flavors the number of ways you can do this is $C(10+30,10) = C(40,10)$ = ${{40!}\over{30! \times 10!}} = 847,660,528$

* Ask the user for an integer value and assign it to r.
* Use loops to compute C(r+30,r) and print the result.
* Convert the final vlaue to a string and assign it to `text`
* Use a for loop with `text` to compute and then print the number of odd and even digits in the string.  (For example, if `text` is "12113305" you would print that there are two even and 6 odd digits.)

* Run this code for at least 5 positive integer values up to at least 1000.  Put the results into a markdown cell.

In a new Markdown section report on the following:
* Run this code for at least 5 positive integer values.  Put the results into a markdown cell.
* How large a value of r can you use before you experience computational issues?  What is the problem python reports?
* As r increases, how does the final result change?
* Looking at your final result values, are the digits more likely to be odd or even?


In [68]:
r = input("Give an integer: ")
r = int(r)
top = 1
bottom = 1
if r > 1 :
    #top loop
    for x in range(1, r + 31) :
        top = top * x
    #bottom loop part 1
    for x in range(1, 31) :
        bottom = bottom * x
    #bottom loop part 2
    for x in range(1, r) :
        bottom = bottom * x
finalvalue = top // bottom
print(finalvalue)

text = str(finalvalue)
even = 0
odd = 0
for i in text :
    if int(i) % 2 == 0 :
        even += 1
    else :
        odd += 1
print("There are " + str(even) + " even numbers and " + str(odd) + " odd numbers.")


Give an integer:  10


8476605280
There are 8 even numbers and 2 odd numbers.


Final Results
* First Test: 1000
    * Answer is 5973989742093443720878720594321533293977443790394706271616000
    * There are 27 even numbers and 34 odd numbers.
* Second Test: 2000
    * Anser is 10203140174768624006527361783882302425030314946845747810369459527212000
    * There are 42 even numbers and 29 odd numbers.
* Third Test: 5000
    * Answer is 19262718176173520955387226532627872014400590660754062885193590219025439815465380000
    * There are 44 even numbers and 39 odd numbers.
* Fourth Test: 10000
    * Answer is 39492453772068981571413857068649806113480741474469402804996383775023606703129183270080260000
    * There are 53 even numbers and 39 odd numbers.
* Fifth Test: 50000
    * Answer is 177193659357861500123119566262505967069090592375273746076306303319841024412105612845674409227529586191989506300000
    * There are 55 even numbers and 59 odd numbers.
* When I try to go up to 1 million the program won't finish on me after waiting for about 10 minutes.
* As r is increasing the final value is increasing very fast.
* 3 out of the 5 times there were more even numbers than odd numbers so I would say that there is a higher chance for a digit to be even.
